src/caffe/datum_imagesdb.cpp

```diff
@@ -0,0 +1,175 @@
+#include <algorithm>
+#include <map>
+#include <string>
+#include <vector>
+
+#include <fstream>  // NOLINT(readability/streams)
+
+#include "caffe/datum_imagesdb.hpp"
+#include "caffe/util/io.hpp"
+#include "caffe/util/rng.hpp"
+
+namespace caffe {
+
+void DatumImagesDB::Open() {
+  if (*is_opened_ == false) {
+    const string& source = param_.source();
+    LOG(INFO) << "Opening imagesdb " << source;
+    datum_database_.reset(new map<string, Datum>());
+    switch (param_.mode()) {
+    case DatumDBParameter_Mode_NEW: {
+      file_.open(source.c_str(), ios::out | ios::trunc);
+      CHECK(file_.good()) << "Could not create file: " << source;
+      keys_.clear();
+      batch_.clear();
+      break;
+    }
+    case DatumDBParameter_Mode_WRITE: {
```

Comment:
We don't actually use `WRITE` mode for anything, do we? I'm not saying it should go away, just something to think about.